File: python/related.py
from timing import finish, lap

TOPN = 5

lap()
from collections import defaultdict

import ujson as json


def main():
    lap()
    with open("../posts.json") as f:
        posts = json.load(f)
    lap()

    tag_map = defaultdict(list)
    for idx, post in enumerate(posts):
        for tag in post["tags"]:
            tag_map[tag].append(idx)

    all_related_posts = [None] * len(posts)
    tagged_post_count = [0] * len(posts)

    for this_post_idx, post in enumerate(posts):
        tagged_post_count = [0] * len(posts)

        for tag in post["tags"]:
            for other_idx in tag_map[tag]:
                tagged_post_count[other_idx] += 1

        tagged_post_count[this_post_idx] = 0

        top5 = [0] * TOPN * 2  # flattened list of (count, id)
        min_tags = 0

        for j, count in enumerate(tagged_post_count):
            if count > min_tags:
                upper_bound = (TOPN - 2) * 2

                while upper_bound >= 0 and count > top5[upper_bound]:
                    top5[upper_bound + 2] = top5[upper_bound]
                    top5[upper_bound + 3] = top5[upper_bound + 1]
                    upper_bound -= 2

                insert_pos = upper_bound + 2
                top5[insert_pos] = count
                top5[insert_pos + 1] = j

                min_tags = top5[TOPN * 2 - 2]

        # faster than list comprehension
        top_posts = [None] * TOPN

        for i in range(1, 10, 2):
            top_posts[i // 2] = posts[top5[i]]

        all_related_posts[this_post_idx] = {
            "_id": post["_id"],
            "tags": post["tags"],
            "related": top_posts,
        }

    lap()
    with open("../related_posts_python.json", "w") as f:
        json.dump(all_related_posts, f)
    lap()
    finish()


if __name__ == "__main__":
    main()

File: nim/related.nim
import std/hashes
import pkg/[jsony, xxhash]
import ./related/fixedtable

when defined(profileGen):
  echo "Generating profile"
else:
  import std/[monotimes, times]

type
  Post = ref object
    `"_id"`: string
    title: string
    tags : seq[string]

  RelatedPosts = object
    `"_id"`: string
    tags : seq[string]
    related : seq[Post]

const
  input = "../posts.json"
  output = "../related_posts_nim.json"

func hash(x: string): Hash =
  cast[Hash](XXH3_64bits(x))

func findTop5(
    posts: seq[Post],
    taggedPostCount: seq[uint8]): seq[Post] =
  result = newSeq[Post](5)
  var top5: array[5, tuple[idx: int, count: uint8]]
  for i, count in taggedPostCount:
    if count > top5[4].count:
      top5[4] = (idx: i, count: count)
      for pos in countdown(3, 0):
        if count > top5[pos].count:
          swap(top5[pos+1], top5[pos])
  for i, t in top5:
    result[i] = posts[t.idx]

func genTagMap(posts: seq[Post]): Table[string, seq[int]] =
  result = initTable[string, seq[int]](100)
  for i, post in posts:
    for tag in post.tags:
      result.withValue(tag, val):
        val[].add i
      do:
        result[tag] = @[i]

proc countTaggedPost(
    taggedPostCount: var seq[uint8],
    i: int,
    post: Post,
    tagMap: Table[string, seq[int]]) =
  for tag in post.tags:
    for otherIDX in tagMap[tag]:
      inc(taggedPostCount[otherIDX])
  taggedPostCount[i] = 0 # remove self

func process(posts: seq[Post]): seq[RelatedPosts] =
  result = newSeq[RelatedPosts](posts.len)
  let tagMap = genTagMap(posts)
  for i, post in posts.pairs:
    var taggedPostCount = newSeq[uint8](posts.len)
    taggedPostCount.countTaggedPost(i, post, tagMap)
    let relatedPosts = RelatedPosts(
      `"_id"`: post.`"_id"`,
      tags: post.tags,
      related: posts.findTop5(taggedPostCount))
    result[i] = relatedPosts

proc main() =
  let posts = readFile(input).fromJson(seq[Post])
  when not defined(profileGen):
    let t0 = getMonotime()
  let res = process(posts)
  when not defined(profileGen):
    let time = (getMonotime() - t0).inMicroseconds / 1000
    echo "Processing time (w/o IO): ", time, "ms"
  writeFile(output, res.toJson)

when isMainModule:
  main()
